src/map/status.cpp

```diff
@@ -9258,6 +9265,16 @@ int status_change_start(struct block_list* src, struct block_list* bl,enum sc_ty
 
 	// Check failing SCs from list
 	if (!scdb->fail.empty()) {
+		// Don't let OPT1 that have RemoveOnDamaged start a new effect
+		if (sc->opt1 > OPT1_NONE && scdb->flag[SCF_REMOVEONDAMAGED]) {
+			unit_data *ud = unit_bl2ud(bl);
+
+			if (ud != nullptr && ud->lastEffect == type) {
+				ud->lastEffect = SC_NONE;
+				return 0;
```

Comment:
This code seems weird to me. You don't actually compare lastEffect against the fail list here. Also what does "RemoveOnDamaged" of the new SC have to do with the skill failing?
And "lastEffect" should also be reset if it does not lead to failing.

For example if you attack a frozen target with 100% sleep attack. Then sleep should still happen because freeze does not make sleep fail. But attacking a frozen target with a 100% petrify attack should just remove freeze but not petrify the target. Whiteimprison also should not block itself from being reapplied.

Basically you don't want to check for OPT1 or RemoveOnDamaged here at all, instead you want to put this logic inside the scdb->fail iteration and just check for "it == ud->lastEffect" (not sure about syntax) as fail condition.